### ISSUE TITLE
Support for putting debug build log output to file

### ISSUE THREAD

**Issue by staffanf** [gradle/gradle#1005]
It would be nice to be able to put debug level build log output into a file while retaining info level build log output in the console.
That way it is possible to peek into the log file when the console output indicates a problem.

### Expected Behavior
I would like to be able to configure logging for the gradle build.
Preferrably to be able to configure destinations (file/console/other?) and thresholds from build scripts.

### Current Behavior
There is no way to configure this right now.

### Context
At times when running regression test on jenkins, there are intermittent problems/tests. Having more complete build/test logs on disk would simplify debugging when reproducing is not always easy or is very time consuming.

This issue was https://issues.gradle.org/browse/GRADLE-1196 with 15 votes.

**Comment by eriwen** [comment]
@staffanf would you consider it necessary to have fine-grained control over which log levels are logged where? For example: what if there were a flag to redirect all logs elsewhere yet keep LIFECYCLE, WARN, and ERROR in the console? Would that solve all, some, or none of your goals?

**Comment by RupprechJo** [comment]
I'm not the bug author, but we want this feature for years! It would be good, to have a seperate log level for console and for file, normally I want more informations in a log file, than I want to see on the console, the user uses. It is not onyl usefull for tests on jenkins, it is also usefull if you have a lot of users using own plugins, and they have problems, they can send you the log file with much information in it to reproduce the bug

**Comment by staffanf** [comment]
@eriwen I think some amount of control would be good. Personally I'm not sure I want debug level logging to file, I think it adds a bit much internal gradle stuff to the logs. Not sure though, since that might be whats needed to fix a some problems.
If I had the configuration option I just might start with both an info and a debug log written to disk. Then trouble shooting by checking the info log first and then going to the debug log if needed.
SIDENOTE:
I kinda always missed a log level between LIFECYCLE and INFO since starting to use gradle. The idea behind that being that if I develop custom tasks or write doLast actions, there is no log level where I can output additional info without also adding **a lot**  of internal gradle logging. Haven't seen any calls for this though, so maybe it's just me.

**Comment by eriwen** [comment]
> it is also usefull if you have a lot of users using own plugins, and they have problems, they can send you the log file with much information in it to reproduce the bug

@RupprechJo This is _exactly_ one of the things we're trying to solve with [build scans](https://scans.gradle.com/s/pgjgssca2kkli), especially because it has the logs but also much more context like environment, dependency, and customizable information.

@staffanf Interesting. I've wanted the same thing but have always survived with INFO logging. My sentiment is: I want to debug just my build and not Gradle.

**Comment by RupprechJo** [comment]
@eriwen I had a quick look at build scan and I think it is more a performance optimizing tool. What I mean: I come from the backend world. In EVERY good application you have a lot more logging, than the normal user sees. And this is good for analyzing strange bugs that happen only once or are hardly to reproduce. 

So for gradle we supply plugins for our enterprise developers. For them the normal console logging is the user interface and this should be nice and clean. But if something went wrong, I need the log for exactly that build that has just run. And it has to contain much more information than the console logging. We automize things with gradle, which are not always idempotent. So trying it another time does not necessarily get the same outcome. 

And as staffanf mentioned, for intensive automated testing it is also very helpful to have the log of exactly that build run.

**Comment by staffanf** [comment]
@eriwen Well I really agree with that sentiment. That tells me that there is at room for improvement regarding logging in gradle e.g. if I could filter packages/classes/loggers like in log4j then I could create a cleaner debug log with info only for my custom packages.
@RupprechJo  I agree with general concept of clean console and useful debug logs.

**Comment by darrenoakey** [comment]
To me - there are two several warring priorities going on here
- we want the cleanest possible build output
- we want to enough information to troubleshoot a problem

I think the key thing to observe is that we never want to see information _unless_ there's a problem.   If the compile task succeeds, I want to see the word "compile" in my CI logs.  If it fails - I want every tiny little bit of information possible there.

The - there's the does the system _really know_ if there's a problem.  ie - a task might succeed, and there still might be a problem, so also - I want to be able _to find out_

All of which seems to me to suggest a logging strategy

- if task succeeds, task output is minimal (ie name only)  
- if task fails - task output is at info level... (ie everything from within the task, but not gradle stuff)
- and ALWAYS.... there's a file on the disk - ideally something like 20170401_19241232_{task_name}_gradle_debug.log - which altogether contain the complete gradle debug log... so that in that one in a thousand times I actually need to go a little deeper into a build- I can - and I can do it easily from a CI system by clicking on the one task I need to - without having to download the entire debug log.

------
I think the key thing to keep in mind is that in many situations... well - at least in mine - no human _ever_ runs gradle.  This only happens in the build farm.  You aren't really in a position to investigate easily - and you certainly aren't in a position to change parameters and run again.   We want to know what _happened_ in the past.   

However - against that - you want to be able to read the build log.  We have it set to info because when we need the information, we _need_ it... but this means we are clicking in jenkins and downloading a 36mb build log (literally) - when all we really want to know is - "test blahblahblah failed because of a null pointer exception on line 36"   or    "you violated this pmd error on this line of code".     

In the perfect world, build output from a test failure would be:
:blah:compileJava
:blah:processResources
:blah:test   FAILED
       test xyx failed
             some stack trace
             any stdout/stderr from that test

because fundamentally....   _no other information is useful_

**Comment by eatdrinksleepcode** [comment]
Has there been any movement on this? The ability to keep the build output clean but still have access to low-level logging if something goes wrong is an important need. Using things like log4j in our own apps, and then trying to debug Gradle builds where we do not have this capability is very frustrating.

**Comment by SimonLammer** [comment]
Are there any updates on this? It would be very helpful for me, not having to dump all information on the console, but instead only a minimum while the full log can still be accessed by looking at a file.

**Comment by Jonatha1983** [comment]
Hi 

I think this is a must. 

1) The scan does not always works, for instance: composite build 
2) Sometime you want the log after a user ran the build (without scan) 

Thanks

**Comment by eriwen** [comment]
@Jonatha1983 I agree there are some good use cases for this, but I want to point out that 1) you can use build scans with composite builds now, and 2) you can use `gradle buildScanPublishPrevious` to capture a build scan for the most recent build after it's run. See https://docs.gradle.com/build-scan-plugin/#publishing_on_demand :)

**Comment by Jonatha1983** [comment]
Hi @eriwen 
 
We are currently using version 4.10.2 and when running with '--scan' (in a composite build project)  Gradle print a message that says something like: "scan is not supported in a composite build" ( and at the end of the run , it does not give any link )  did you mean that in 5.0 it will work with composite build ? or maybe does it work  for the enterprise only?
 
Also my second point was not for the most recent build - it is a common scenario where a developer is doing something wrong calling Devops/System engineer for help and in turn he/she need to investigate the flow history. 

I found this blog:
[http://willis7.github.io/blog/2013/gradle-output-to-log.html](url)
It  make use of some Gradle's internal packages to collect execution log only
I had to change the import statement but it worked.

I think Gradle has already all the infrastructure to give the user a simple way to configure the output to a log file - really hope to see it in some next release.

**Comment by szpak** [comment]
I think, there are use cases where putting a debug log to a file while preserving the regular console output might be useful.

**Comment by lwasyl** [comment]
Another use case: publishing build scan will fail with `The build scan publishing request was rejected as the data exceeded the maximum size.` if the output is too long. So right now it's not possible to run `./gradlew build -Dorg.gradle.caching.debug=true --scan`, because caching debug logs are massive and cause the build scan to be rejected due to size. If it was possible to redirect caching debug logs to a file, the scan would succeed

**Comment by AndrewShinCoder** [comment]
Publishing build.scan will always reject the following code:


./grawdweel{build -null}

**Comment by gallyamb** [comment]
My use case here is to output debug logs into a file and then put that file into an artifact of the job in the GitLab CI, as GitLab CI job's logs are restricted by their size. I'm trying to find the root cause of the Configuring stage stuck at 100%, and there is no way to redirect debug logs into a file while keeping standard logs in STDOUT
